### PR TITLE
ATO-580: Added Client Registration Lambda

### PR DESCRIPTION
**Pull request by BenjaminWCO**
## What

Added Client Registration Handler

## How to review

Code Review

## Checklist

- [x] Impact on orch and auth mutual dependencies has been checked.

### CODE REVIEW

File: ci/terraform/oidc/api-gateway.tf [hunk 1]
```diff
@@ -946,3 +948,39 @@ resource "aws_api_gateway_integration" "orch_token_integration" {
...
+    module.register
+  ]
+}
+
+resource "aws_api_gateway_method" "orch_register_method" {
+  count       = var.orch_register_enabled ? 1 : 0
+  rest_api_id = aws_api_gateway_rest_api.di_authentication_api.id
+  resource_id = aws_api_gateway_resource.orch_register_resource[0].id
+  http_method = "POST"
+  api_key_required   = true;
```
Comment (CarlyG55) [inline ci/terraform/oidc/api-gateway.tf:967]: no semi colon 🙂

File: template.yaml [hunk 2]
```diff
@@ -842,6 +842,198 @@ Resources:
...
+          TXMA_AUDIT_QUEUE_URL:
+            Fn::ImportValue: !Sub orchestration-${Environment}-txma-QueueURL
+          REDIS_KEY: session
+          INTERNAl_SECTOR_URI: !Sub
+            - https://identity.${ServiceDomain}/
+            - ServiceDomain: !FindInMap [ EnvironmentConfiguration, !Ref Environment, serviceDomain ]
+          DYNAMO_ARN_PREFIX: !Sub
+            - 'arn:aws:dynamodb:eu-west-2:${AccountId}:table/${AuthEnvironment}-'
+            - AccountId: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authAccountId ]
+              AuthEnvironment: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authEnvironment ]
```
Comment (CarlyG55) [inline template.yaml:872]:
```suggestion
          ENVIRONMENT: !Sub ${Environment}
          DYNAMO_ARN_PREFIX: !Sub
            - 'arn:aws:dynamodb:eu-west-2:${AccountId}:table/${AuthEnvironment}-'
            - AccountId: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authAccountId ]
              AuthEnvironment: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authEnvironment ]
```
all of the rest of these aren't needed

File: template.yaml [hunk 3]
```diff
@@ -842,6 +842,198 @@ Resources:
...
+          - Sid: AllowUserProfileTableKms
+            Effect: Allow
+            Action:
+              - kms:Encrypt
+              - kms:Decrypt
+              - kms:ReEncrypt*
+              - kms:GenerateDataKey*
+              - kms:CreateGrant
+              - kms:DescribeKey
+            Resource: !FindInMap [ EnvironmentConfiguration, !Ref Environment, userProfileTableKeyArn ]
```
Comment (CarlyG55) [inline template.yaml:947]:
```suggestion
          - Sid: AllowClientRegistryTableReadWriteAccess
            Effect: Allow
            Action:
              - dynamodb:BatchGetItem
              - dynamodb:DescribeTable
              - dynamodb:Get*
              - dynamodb:Query
              - dynamodb:Scan
              - dynamodb:UpdateItem
              - dynamodb:PutItem
            Resource: !Sub
              - 'arn:aws:dynamodb:eu-west-2:${AccountId}:table/${AuthEnvironment}-client-registry'
              - AccountId: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authAccountId ]
                AuthEnvironment: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authEnvironment ]
          - Sid: AllowClientRegistryTableKms
            Effect: Allow
            Action:
              - kms:Encrypt
              - kms:Decrypt
              - kms:ReEncrypt*
              - kms:GenerateDataKey*
              - kms:CreateGrant
              - kms:DescribeKey
            Resource: !FindInMap [ EnvironmentConfiguration, !Ref Environment, clientRegistryTableKeyArn ]
          - Sid: AllowTxMAEventKms
            Effect: Allow
            Action:
              - kms:Encrypt
              - kms:Decrypt
              - kms:ReEncrypt*
              - kms:GenerateDataKey*
              - kms:CreateGrant
              - kms:DescribeKey
            Resource: !FindInMap [ EnvironmentConfiguration, !Ref Environment, txmaEventEncryptionKeyArn ]
          - Sid: AllowSQSTxMA
            Effect: Allow
            Action:
              - sqs:SendMessage
            Resource:
              - !Sub arn:aws:sqs:eu-west-2:${AWS::AccountId}:orchestration-${Environment}-txma-AuditEventQueue
```
don't need a bunch of these permissions, and need to add the txma ones 🙂

File: ci/terraform/oidc/api-gateway.tf (hunk 1 above)
Comment (BenjaminWCO) [reply ci/terraform/oidc/api-gateway.tf]: fixed

File: template.yaml (hunk 2 above)
Comment (BenjaminWCO) [reply template.yaml]: Thanks, not sure I was doing. I think it needs txma_audit_queue.

File: template.yaml [hunk 4]
```diff
@@ -1315,6 +1315,159 @@ Resources:
...
             Unit: Count
 
+  ClientRegistrationFunction:
+    Type: AWS::Serverless::Function
+    # checkov:skip=CKV_AWS_116: DLQ is not appropriate for a Lambda invoked by an API
+    # checkov:skip=CKV_AWS_117: VPC settings are global
+    Properties:
+      FunctionName: !Sub ${Environment}-ClientRegistrationFunction
+      AutoPublishAlias: latest
+      CodeUri: ./oidc-api
```
Comment (CarlyG55) [inline template.yaml:1325]:
```suggestion
      CodeUri: ./client-registry-api
```

File: template.yaml [hunk 5]
```diff
@@ -1315,6 +1315,159 @@ Resources:
...
+    Type: AWS::Logs::LogGroup
+    Properties:
+      LogGroupName: !Sub /aws/lambda/${Environment}-register-lambda
+      KmsKeyId: !GetAtt MainKmsKey.Arn
+      RetentionInDays: 14
+
+  ClientRegistrationFunctionErrorMetricFilter:
+    Type: AWS::Logs::MetricFilter
+    Properties:
+      FilterName: !Sub ${Environment}-register-logout-request-errors
```
Comment (CarlyG55) [inline template.yaml:1406]:
```suggestion
      FilterName: !Sub ${Environment}-register-request-errors
```

**Review by CarlyG55 — DISMISSED** [review]
2 tiny things, happy with it once those are fixed (I might need to reapprove though so feel free to ping me if so)

**Review by CarlyG55 — APPROVED** [review]

**Review by ethanmills — APPROVED** [review]

**Review by ethanmills — APPROVED** [review]

**Pull request merged**